File: core/network.py
"""
网络工具模块
包含所有网络相关的操作函数
"""
import subprocess
import re
import socket


def get_interfaces():
    """获取所有网卡及其IP地址"""
    output = subprocess.check_output(
        "ipconfig /all",
        shell=True,
        encoding="gbk",
        errors="ignore"
    )

    interfaces = []
    current_iface = None
    current_ip = None

    for line in output.splitlines():
        line = line.strip()

        name_match = re.match(
            r'(以太网适配器|无线局域网适配器|Ethernet adapter|Wireless LAN adapter)\s+(.+?):',
            line
        )
        if name_match:
            if current_iface:
                interfaces.append((current_iface, current_ip or "未获取"))
            current_iface = name_match.group(2)
            current_ip = None
            continue

        if current_iface:
            ip_match = re.search(
                r'(IPv4 地址|IPv4 Address|IP Address)[^0-9]*(\d+\.\d+\.\d+\.\d+)',
                line
            )
            if ip_match:
                current_ip = ip_match.group(2)

    if current_iface:
        interfaces.append((current_iface, current_ip or "未获取"))

    return interfaces


def run(cmd):
    """执行系统命令"""
    subprocess.run(cmd, shell=True, check=True)


def set_static_ip(iface, ip, mask):
    """设置静态IP地址"""
    run(f'netsh interface ipv4 set address name="{iface}" static {ip} {mask}')


def set_dns(iface, dns):
    """设置DNS服务器"""
    run(f'netsh interface ipv4 set dns name="{iface}" static {dns}')


def set_mtu(iface, mtu):
    """设置MTU值"""
    run(f'netsh interface ipv4 set subinterface "{iface}" mtu={mtu} store=persistent')


def add_route(gateway):
    """添加永久路由"""
    run(f'route -p add 10.0.0.0 mask 255.0.0.0 {gateway}')


def set_all_mtu(mtu=1300):
    """设置所有网卡的MTU"""
    interfaces = get_interfaces()
    results = []
    for name, ip in interfaces:
        try:
            set_mtu(name, mtu)
            results.append(f"✓ {name} MTU已设置为 {mtu}")
        except Exception as e:
            results.append(f"✗ {name} 设置失败: {str(e)}")
    return results


def get_default_gateway():
    """从路由表中获取默认网关IP（0.0.0.0 mask 0.0.0.0）"""
    try:
        output = subprocess.check_output(
            'route print -4',
            shell=True,
            encoding='gbk',
            errors='ignore'
        )
        
        # 查找默认路由（0.0.0.0 mask 0.0.0.0）
        for line in output.splitlines():
            line = line.strip()
            if '0.0.0.0' in line and '0.0.0.0' in line:
                parts = line.split()
                if len(parts) >= 3:
                    for i, part in enumerate(parts):
                        if part == '0.0.0.0' and i > 0:
                            if i + 1 < len(parts) and parts[i + 1] == '0.0.0.0':
                                gateway = parts[i + 2] if i + 2 < len(parts) else None
                                if gateway and gateway != 'On-link':
                                    return gateway
        return None
    except:
        return None


def ping_host(host, count=4):
    """ping指定主机，返回是否成功和结果信息"""
    try:
        # Windows系统使用 -n 参数
        output = subprocess.check_output(
            f"ping -n {count} {host}",
            shell=True,
            encoding="gbk",
            errors="ignore",
            timeout=10
        )
        # 检查是否有成功的响应
        if "TTL=" in output or "ms" in output:
            # 提取统计信息
            if "已发送 = " in output:
                # 中文版本
                match = re.search(r'已发送 = (\d+)，已接收 = (\d+)，丢失 = (\d+)', output)
                if match:
                    sent, received, lost = match.groups()
                    return True, f"已发送 {sent} 个，已接收 {received} 个，丢失 {lost} 个"
            elif "Sent = " in output:
                # 英文版本
                match = re.search(r'Sent = (\d+), Received = (\d+), Lost = (\d+)', output)
                if match:
                    sent, received, lost = match.groups()
                    return True, f"Sent {sent}, Received {received}, Lost {lost}"
            return True, "ping 成功"
        return False, "ping 失败，无响应"
    except subprocess.TimeoutExpired:
        return False, "ping 超时"
    except Exception as e:
        return False, f"ping 失败: {str(e)}"


def test_host_connectivity(host, port=80, timeout=3):
    """测试主机端口连通性"""
    try:
        socket.create_connection((host, port), timeout=timeout)
        return True
    except:
        return False


def route_already_set():
    """检查路由是否已配置"""
    try:
        output = subprocess.check_output(
            'route print 10.0.0.0',
            shell=True,
            encoding='gbk',
            errors='ignore'
        )
        return '10.0.0.0' in output
    except:
        return False


def ip_already_set(iface):
    """检查指定网卡的IP是否已配置为10.36.x.x"""
    for name, ip in get_interfaces():
        if name == iface and ip.startswith("10.36."):
            return True
    return False


def mtu_already_set(iface):
    """检查指定网卡的MTU是否已设置为1300"""
    try:
        output = subprocess.check_output(
            f'netsh interface ipv4 show interface "{iface}"',
            shell=True,
            encoding='gbk',
            errors='ignore'
        )
        return bool(re.search(r'MTU\s*:\s*1300', output))
    except:
        return False

File: internal/network/network.go
// Package network 负责网卡解析、连通性测试与 netsh/route 网络配置
package network

import (
	"fmt"
	"net"
	"regexp"
	"strconv"
	"strings"
	"time"

	"gnetconf/internal/config"
	"gnetconf/internal/hosts"
	"gnetconf/internal/system"
)

// InterfaceInfo 描述一个网卡及其 IPv4 地址
type InterfaceInfo struct {
	Name string
	IP   string
}

var (
	adapterRe     = regexp.MustCompile(`(?:以太网适配器|无线局域网适配器|Ethernet adapter|Wireless LAN adapter)\s+(.+?):`)
	ipRe          = regexp.MustCompile(`(?:IPv4 地址|IPv4 Address|IP Address)[^0-9]*(\d+\.\d+\.\d+\.\d+)`)
	rePingSent    = regexp.MustCompile(`已发送 = (\d+)`)
	rePingLost    = regexp.MustCompile(`丢失 = (\d+)`)
	rePingLossPct = regexp.MustCompile(`(\d+)% 丢失`)
	rePingAvg     = regexp.MustCompile(`平均 = (\d+)ms`)
	reMtu         = regexp.MustCompile(`MTU\s*:\s*1300`)
)

var (
	ifaceCache     []InterfaceInfo
	ifaceCacheTime time.Time
)

const ifaceCacheTTL = 3 * time.Second

// runUTF8 以 UTF-8 代码页执行命令并返回输出（适配中文 Windows 的 GBK 输出）
func runUTF8(line string) (string, error) {
	return system.RunLine(`chcp 65001 >nul && ` + line)
}

// GetInterfaces 解析 ipconfig /all 获取网卡与 IPv4，结果缓存 3 秒避免重复执行系统命令
func GetInterfaces() []InterfaceInfo {
	if time.Since(ifaceCacheTime) < ifaceCacheTTL && ifaceCache != nil {
		return ifaceCache
	}
	out, err := runUTF8(`ipconfig /all`)
	if err != nil && out == "" {
		return nil
	}
	var result []InterfaceInfo
	var curName, curIP string
	for _, line := range strings.Split(out, "\n") {
		line = strings.TrimSpace(line)
		if m := adapterRe.FindStringSubmatch(line); m != nil {
			if curName != "" {
				result = append(result, InterfaceInfo{Name: curName, IP: curIP})
			}
			curName = strings.TrimSpace(m[1])
			curIP = "未获取"
			continue
		}
		if curName != "" {
			if m := ipRe.FindStringSubmatch(line); m != nil {
				curIP = m[1]
			}
		}
	}
	if curName != "" {
		result = append(result, InterfaceInfo{Name: curName, IP: curIP})
	}
	ifaceCache = result
	ifaceCacheTime = time.Now()
	return result
}

// RefreshInterfaces 清除网卡缓存，下次调用 GetInterfaces 时重新获取
func RefreshInterfaces() {
	ifaceCacheTime = time.Time{}
}

// PingResult 一次 ping 的结果
type PingResult struct {
	Host      string
	Sent      int
	Lost      int
	LossPct   int
	AvgTimeMs int
	OK        bool
}

// Ping 对目标执行 ping 测试
func Ping(host string, count, timeoutMs int) PingResult {
	r := PingResult{Host: host}
	out, err := system.RunLine(fmt.Sprintf(`ping -n %d -w %d %s`, count, timeoutMs, host))
	if err != nil && out == "" {
		return r
	}
	if m := rePingSent.FindStringSubmatch(out); m != nil {
		r.Sent, _ = strconv.Atoi(m[1])
	}
	if m := rePingLost.FindStringSubmatch(out); m != nil {
		r.Lost, _ = strconv.Atoi(m[1])
	}
	if m := rePingLossPct.FindStringSubmatch(out); m != nil {
		r.LossPct, _ = strconv.Atoi(m[1])
	}
	if m := rePingAvg.FindStringSubmatch(out); m != nil {
		r.AvgTimeMs, _ = strconv.Atoi(m[1])
	}
	r.OK = r.Lost < r.Sent && r.Sent > 0
	return r
}

// TestHostConnectivity TCP 连通性测试
func TestHostConnectivity(host string, port, timeoutSec int) bool {
	addr := net.JoinHostPort(host, strconv.Itoa(port))
	conn, err := net.DialTimeout("tcp", addr, time.Duration(timeoutSec)*time.Second)
	if err != nil {
		return false
	}
	_ = conn.Close()
	return true
}

// SetStaticIP 设置静态 IP 与子网掩码
func SetStaticIP(iface, ip, mask string) error {
	err := system.Run("netsh", "interface", "ipv4", "set", "address",
		fmt.Sprintf(`name=%q`, iface), "static", ip, mask)
	RefreshInterfaces()
	return err
}

// SetDNS 设置 DNS
func SetDNS(iface, dns string) error {
	err := system.Run("netsh", "interface", "ipv4", "set", "dns",
		fmt.Sprintf(`name=%q`, iface), "static", dns)
	RefreshInterfaces()
	return err
}

// SetMTU 设置网卡 MTU
func SetMTU(iface string, mtu int) error {
	err := system.Run("netsh", "interface", "ipv4", "set", "subinterface",
		fmt.Sprintf(`"%s"`, iface), fmt.Sprintf("mtu=%d", mtu), "store=persistent")
	RefreshInterfaces()
	return err
}

// AddRoute 添加永久路由
func AddRoute(gateway string) error {
	err := system.Run("route", "-p", "add", config.DefaultRoute, "mask", config.DefaultRouteMask, gateway)
	RefreshInterfaces()
	return err
}

// SetAllMTU 设置所有网卡 MTU
func SetAllMTU(mtu int) []string {
	var results []string
	for _, i := range GetInterfaces() {
		if err := SetMTU(i.Name, mtu); err != nil {
			results = append(results, fmt.Sprintf("✗ %s 设置失败: %v", i.Name, err))
		} else {
			results = append(results, fmt.Sprintf("✓ %s MTU已设置为 %d", i.Name, mtu))
		}
	}
	return results
}

// GetDefaultGateway 从路由表获取默认网关
func GetDefaultGateway() string {
	out, err := runUTF8(`route print -4`)
	if err != nil && out == "" {
		return ""
	}
	for _, line := range strings.Split(out, "\n") {
		line = strings.TrimSpace(line)
		if strings.Count(line, "0.0.0.0") >= 2 {
			fields := strings.Fields(line)
			for i, f := range fields {
				if f == "0.0.0.0" && i+2 < len(fields) && fields[i+1] == "0.0.0.0" {
					gw := fields[i+2]
					if gw != "On-link" {
						return gw
					}
				}
			}
		}
	}
	return ""
}

// IpAlreadySet 判断网卡是否已配置 10.36. 段 IP
func IpAlreadySet(iface string) bool {
	for _, i := range GetInterfaces() {
		if i.Name == iface && strings.HasPrefix(i.IP, config.DefaultIPPrefix) {
			return true
		}
	}
	return false
}

// MtuAlreadySet 判断网卡 MTU 是否为 1300
func MtuAlreadySet(iface string) bool {
	out, err := runUTF8(fmt.Sprintf(`netsh interface ipv4 show interface "%s"`, iface))
	if err != nil && out == "" {
		return false
	}
	return reMtu.MatchString(out)
}

// RouteAlreadySet 判断 10.0.0.0 路由是否已存在
func RouteAlreadySet() bool {
	out, err := runUTF8(`route print ` + config.DefaultRoute)
	if err != nil && out == "" {
		return false
	}
	return strings.Contains(out, config.DefaultRoute)
}

// GetMissingItems 返回指定网卡缺失的配置项
func GetMissingItems(iface string) []string {
	var missing []string
	if !IpAlreadySet(iface) {
		missing = append(missing, "IP 地址")
	}
	if !MtuAlreadySet(iface) {
		missing = append(missing, "MTU")
	}
	if !hosts.AlreadySet() {
		missing = append(missing, "hosts 文件")
	}
	if !RouteAlreadySet() {
		missing = append(missing, "路由")
	}
	return missing
}

// ProgressCB 配置进度回调
type ProgressCB func(current, total int, message string)

// ApplyMissingConfig 按缺失项应用配置
func ApplyMissingConfig(iface, ip, mask, dns string, missing []string, cb ProgressCB) {
	total := len(missing)
	cur := 0
	step := func(msg string) {
		cur++
		if cb != nil {
			cb(cur, total, msg)
		}
	}

	if system.StringsContains(missing, "IP 地址") {
		if err := SetStaticIP(iface, ip, mask); err != nil {
			system.Trace("SetStaticIP 失败: " + err.Error())
		}
		if err := SetDNS(iface, dns); err != nil {
			system.Trace("SetDNS 失败: " + err.Error())
		}
		step("正在配置IP地址...")
	}
	if system.StringsContains(missing, "路由") {
		gw := strings.Join(strings.Split(ip, ".")[:3], ".") + ".1"
		if err := AddRoute(gw); err != nil {
			system.Trace("AddRoute 失败: " + err.Error())
		}
		step("正在添加路由...")
	}
	if system.StringsContains(missing, "MTU") {
		if err := SetMTU(iface, config.DefaultMTU); err != nil {
			system.Trace("SetMTU 失败: " + err.Error())
		}
		step("正在设置MTU...")
	}
	if system.StringsContains(missing, "hosts 文件") {
		if added, err := hosts.Modify(); err != nil {
			system.Trace("hosts.Modify 失败: " + err.Error())
		} else {
			_ = added
		}
		step("正在修改hosts文件...")
	}
}
